datafusion/expr/src/udf.rs

```diff
@@ -481,15 +563,15 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// This function must consistently return the same type for the same
     /// logical input even if the input is simplified (e.g. it must return the same
     /// value for `('foo' | 'bar')` as it does for ('foobar').
-    fn return_type_from_exprs(
-        &self,
-        _args: &[Expr],
-        _schema: &dyn ExprSchema,
-        arg_types: &[DataType],
-    ) -> Result<DataType> {
-        self.return_type(arg_types)
+    fn return_type_from_args(&self, args: ReturnTypeArgs) -> Result<ReturnInfo> {
+        let return_type = self.return_type(args.arg_types)?;
```

Comment:
This change was a bit problematic for users with custom UDFs in that it may cause breaks at runtime. 

The `#[deprecated()]` apparently doesn't affect for people _implementing_ this trait, only those who call it. As such, if someone (like us) had implemented `return_type_from_exprs` and had `return_type` just panic, that'd compile fine, but rather than calling the `return_type_from_exprs` DF would call this new `return_type_from_args` with its default impl delegating to `return_type` which would then panic.

Luckily, we had tests that caught it, but in general I think in this case it would have been better to fully remove the `return_type_from_exprs` so that users who override it would get compile failures, rather than runtime ones.

Reply:
> Luckily, we had tests that caught it, but in general I think in this case it would have been better to fully remove the return_type_from_exprs so that users who override it would get compile failures, rather than runtime ones.


THis makes sense to me - can you file a ticket? I think this one is related but I am not sure it is entirely the same
- https://github.com/apache/datafusion/issues/14652